### PR TITLE
locks in migrations

### PR DESCRIPTION
**Pull request by jbenet**
This PR adds locking to each migration. I've tested it to work
correctly under these circumstances:

- fs-repo-migrations defers stopping to each migration
- 0-to-1 stops if repo is locked by daemon or cli
- 1-to-2 stops if repo is locked by daemon or cli
- 1-to-2 removes the daemon.lock file if it succeeds
- 1-to-2 locks repo.lock on revert (does not remove it)

### CODE REVIEW

**Comment by jbenet** [comment]
@whyrusleeping im going to remove b0d13e8 from master so this merges cleanly

File: ipfs-0-to-1/migration/migration.go [hunk 1]
```diff
@@ -30,6 +31,12 @@ func (m Migration) Reversible() bool {
 func (m Migration) Apply(opts migrate.Options) error {
 	repo := mfsr.RepoPath(opts.Path)
```
Comment (whyrusleeping) [inline ipfs-0-to-1/migration/migration.go]: why do this before locking?

File: ipfs-1-to-2/go-uuid/.hg/cache/branch2-base [hunk 2]
```diff
@@ -0,0 +1,2 @@
+5fac954758f5aa282478dc07fbb8709fe9575184 9
+5fac954758f5aa282478dc07fbb8709fe9575184 o default
```
Comment (whyrusleeping) [inline ipfs-1-to-2/go-uuid/.hg/cache/branch2-base]: what is all this about?

File: ipfs-1-to-2/go-uuid/.hg/hgrc [hunk 3]
```diff
@@ -0,0 +1,2 @@
+[paths]
```
Comment (whyrusleeping) [inline ipfs-1-to-2/go-uuid/.hg/hgrc]: bad. bad mercurial! youre just an inferior git!

File: ipfs-1-to-2/migration/migration.go [hunk 4]
```diff
@@ -71,10 +87,26 @@ func (m Migration) Apply(opts migrate.Options) error {
...
 }
 
 func (m Migration) Revert(opts migrate.Options) error {
+	repolk, err := lock.Lock2(opts.Path) // lock repo.lock
+	if err != nil {
+		return err
+	}
+	defer repolk.Close()
+
 	repo := mfsr.RepoPath(opts.Path)
```
Comment (whyrusleeping) [inline ipfs-1-to-2/migration/migration.go]: and here you create `repo` after the lock.

**Comment by whyrusleeping** [comment]
A few comments, LGTM otherwise

File: ipfs-1-to-2/go-uuid/.hg/hgrc (hunk 3 above)
Comment (jbenet) [inline ipfs-1-to-2/go-uuid/.hg/hgrc]: removed :disappointed:

File: ipfs-1-to-2/migration/migration.go (hunk 4 above)
Comment (jbenet) [inline ipfs-1-to-2/migration/migration.go]: is that bad? will `repo` create the repository if it isn't there or something?

Comment (jbenet) [inline ipfs-1-to-2/migration/migration.go]: it's just a cast isn't it?

Comment (jbenet) [inline ipfs-1-to-2/migration/migration.go]: ahhh nvm. just saw your other comment

File: ipfs-0-to-1/migration/migration.go (hunk 1 above)
Comment (jbenet) [inline ipfs-0-to-1/migration/migration.go]: no need. fixing.

**Pull request merged**